### DISCUSSION TITLE
Typesafety in xUnit with TheoryData<T>

### DISCUSSION THREAD

**Discussion by giscus** [linkdotnet/Blog.Discussions#95] [General]
# Typesafety in xUnit with TheoryData<T>

I recently discovered this small but very useful utility in xUnit: TheoryData<T>.

https://steven-giesel.com/blogPost/a8aa3385-8829-444a-b269-7ecb38aeaf2f

**Comment by ppasieka** [comment]
It is all great, but you have to remember that xUnit is removing BinarySerialization, therefore not every type will work as a T in TheoryData.
Right now only primitive types are working, and arrays, but not custom classes. For these you have to implement IXUnitSerializable interface. You should check if your tests are actually running in different tests runners (R# vs VS vs console) - you might be surprised.

**Reply by linkdotnet** [reply]
Thanks, I wasn't aware of that.

For everyone who is interested in the details: https://github.com/xunit/xunit/discussions/2506
